File: src/leetcode/python/medium/ValidateBinarySearchTree_98.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class ValidateBinarySearchTree_98(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def validSubTree(node, lf, rg):
            if not node:
                return True
            if lf is not None and lf >= node.val:
                return False
            if rg is not None and node.val >= rg:
                return False
            if not node.left and not node.right:
                return True
            return validSubTree(node.left, lf, node.val) and validSubTree(node.right, node.val, rg)

        if not root or not root.left and not root.right:
            return True
        l, v, r = None, root.val, None
        return validSubTree(root.left, l, root.val) and validSubTree(root.right, root.val, r)

File: src/leetcode/java/ValidateBinarySearchTree_98.java
package leetcode.java;

public class ValidateBinarySearchTree_98 {
  public boolean isValidBST(TreeNode root) {
    if (root == null || (root.left == null && root.right == null)) return true;
    long l = Long.MIN_VALUE;
    int v = root.val;
    long r = Long.MAX_VALUE;
    return validSubTree(root.left, l, v) && validSubTree(root.right, v, r);
  }

  private boolean validSubTree(TreeNode node, long l, long r) {
    if (node == null) return true;
    if (l >= node.val || node.val >= r) return false;
    if (node.left == null && node.right == null) return true;
    return validSubTree(node.left, l, node.val) && validSubTree(node.right, node.val, r);
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}
